hydromt/raster.py

```diff
@@ -2247,6 +2248,245 @@ def tile_window(shape, px):
+
+        def create_folder(path):
+            if not os.path.exists(path):
+                os.makedirs(path)
+
+        if bbox:
+            minx, miny, maxx, maxy = bbox
+        else:
+            minx, miny, maxx, maxy = obj.raster.transform_bounds("EPSG:3857")
+
+        minx, miny = map(
+            max,
+            zip((minx, miny), [-20037508.34] * 2),
+        )
+        maxx, maxy = map(
+            min,
+            zip((maxx, maxy), [20037508.34] * 2),
+        )
+
+        _obj_tr = rasterio.warp.calculate_default_transform(
+            obj.raster.crs,
+            "EPSG:3857",
+            *obj.shape,
+            *obj.raster.bounds,
+        )[0]
+        dres = _obj_tr[0]
+        nzl = int(np.ceil((np.log10((20037508.34 * 2) / (dres * 256)) / np.log10(2))))
+
+        def tile_window(zl, minx, miny, maxx, maxy):
+            # Origin displacement
+            odx = np.floor(abs(-20037508.34 - minx) / dx)
+            ody = np.floor(abs(20037508.34 - maxy) / dx)
+
+            # Set the new origin
+            minx = -20037508.34 + odx * dx
+            maxy = 20037508.34 - ody * dx
+
+            # Create window generator
+            lu = product(
+                np.arange(round(minx, 2), round(maxx, 2), dx),
+                np.arange(round(maxy, 2), round(miny, 2), -dx),
+            )
+            for l, u in lu:
+                col = int(round(odx + (l - minx) / dx, 2))
+                row = int(round(ody + (maxy - u) / dx, 2))
+                bounds = [
+                    l,
+                    u - dx,
+                    l + dx,
+                    u,
+                ]
+                yield Affine(px, 0, l, 0, -px, u), bounds, col, row
+
+        zoom_levels.sort()
+        if zoom_levels[-1] > min_lvl:
+            min_lvl = zoom_levels[-1]
+        zoom_levels = list(range(zoom_levels[0], min_lvl + 1, 1))
+        zoom_levels.reverse()
+
+        create_folder(root)
+
+        vrt_fn = None
+        nodata = self.nodata
+        zls = {}
+        for zl in zoom_levels:
+            dx = (20037508.34 * 2) / (2**zl)
+            px = dx / px_size
+
+            if max(zoom_levels) > nzl:
+                logger.warning(
+                    f"Pixels at zoomlevel {zl} are smaller than the original data"
+                )
+
+            # read data from previous zoomlevel
+            if vrt_fn is not None:
+                obj = xr.open_dataarray(vrt_fn, engine="rasterio").squeeze(
+                    "band", drop=True
+                )
+
+            obj_bounds = obj.raster.bounds
+            obj_res = obj.raster.res[0]
+            obj_shape = obj.raster.shape
+
+            t_r = pyproj.Transformer.from_crs(
+                3857, obj.raster.crs.to_epsg(), always_xy=True
+            )
+
+            sd = join(root, f"{zl}")
+            create_folder(sd)
+            txt_path = join(sd, "filelist.txt")
+            file = open(txt_path, "w")
```

Comment:
You're leaking memory here if an exception is raised during the inner loop because this will never get closed. Please use a context manager.

Reply:
If an exception is raised, the python interpreter stops and the garbage collection is called. Won't that just clean this stuff up? (genuinely asking).  c-like memory leaks are quite rare in python.

Reply:
Depends on if someone in the stack catches your exception, which plenty of libs/applications do. Though I'm guessing it's not quite a memory leak, as much as a dangling file descriptor that's never closed? Unsure how that is handled. But regardless, I think it's best to be more careful, taking the first bit into consideration.

Reply:
I agree it's probably better, but this will most likely not cause any harm. I already indented quite a lot due to the flow of control, so I'd rather not use the contextmanager from an aesthetic standpoint.